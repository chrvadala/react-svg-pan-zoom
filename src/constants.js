export const MODE_IDLE = 'idle';
export const MODE_PANNING = 'panning';
export const MODE_ZOOMING = 'zooming';

export const TOOL_AUTO = 'auto';
export const TOOL_NONE = 'none';
export const TOOL_PAN = 'pan';
export const TOOL_ZOOM_IN = 'zoom-in';
export const TOOL_ZOOM_OUT = 'zoom-out';

export const POSITION_NONE = 'none';
export const POSITION_TOP = 'top';
export const POSITION_RIGHT = 'right';
export const POSITION_BOTTOM = 'bottom';
export const POSITION_LEFT = 'left';

export const ACTION_ZOOM = 'zoom';
export const ACTION_PAN = 'pan';

export const ALIGN_CENTER = 'center';
export const ALIGN_LEFT = 'left';
export const ALIGN_RIGHT = 'right';
export const ALIGN_TOP = 'top';
export const ALIGN_BOTTOM = 'bottom';
