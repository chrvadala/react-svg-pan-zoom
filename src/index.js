import {tipNoViewer} from "./migration-tips";

export {default as ReactSVGPanZoom} from './viewer';
export {default as UncontrolledReactSVGPanZoom} from './uncontrolled-viewer';
export {default as Toolbar} from './ui-toolbar/toolbar';
export {default as ToolbarButton} from './ui-toolbar/toolbar-button';
export {default as IconCursor} from './ui-toolbar/icon-cursor';
export {default as IconPan} from './ui-toolbar/icon-pan';
export {default as IconFit} from './ui-toolbar/icon-fit';
export {default as IconZoomIn} from './ui-toolbar/icon-zoom-in';
export {default as IconZoomOut} from './ui-toolbar/icon-zoom-out';
export {default as Miniature} from './ui-miniature/miniature';
export {setPointOnViewerCenter, reset} from './features/common';
export {pan} from './features/pan';
export {zoom, fitSelection, fitToViewer, zoomOnViewerCenter} from './features/zoom';
export {openMiniature, closeMiniature} from './features/miniature'
export * from './constants';

export const Viewer = () => {
  tipNoViewer()
  return null;
};
