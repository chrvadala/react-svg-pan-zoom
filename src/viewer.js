import React from 'react';
import PropTypes from 'prop-types';
import {toSVG} from 'transformation-matrix';

//events
import eventFactory from './events/event-factory';

//features
import {pan} from './features/pan';
import {
  getDefaultValue,
  setViewerSize,
  setSVGSize,
  setPointOnViewerCenter,
  reset,
  setZoomLevels
} from './features/common';
import {
  onMouseDown,
  onMouseMove,
  onMouseUp,
  onWheel,
  onMouseEnterOrLeave,
  onInterval,
  onDoubleClick
} from './features/interactions';
import {
  onTouchStart,
  onTouchMove,
  onTouchEnd,
  onTouchCancel
} from './features/interactions-touch';

import {zoom, fitSelection, fitToViewer, zoomOnViewerCenter} from './features/zoom';
import {openMiniature, closeMiniature} from './features/miniature';

//ui
import cursorPolyfill from './ui/cursor-polyfill';
import BorderGradient from './ui/border-gradient';
import Selection from './ui/selection';
import Toolbar from './ui-toolbar/toolbar';
import detectTouch from './ui/detect-touch';
import Miniature from './ui-miniature/miniature'

import {
  TOOL_AUTO, TOOL_NONE, TOOL_PAN, TOOL_ZOOM_IN, TOOL_ZOOM_OUT,
  MODE_IDLE, MODE_PANNING, MODE_ZOOMING,
  POSITION_NONE, POSITION_TOP, POSITION_RIGHT, POSITION_BOTTOM, POSITION_LEFT,
  ACTION_PAN, ACTION_ZOOM,
  ALIGN_CENTER, ALIGN_LEFT, ALIGN_RIGHT, ALIGN_TOP, ALIGN_BOTTOM
} from './constants';

export default class ReactSVGPanZoom extends React.Component {

  constructor(props, context) {
    super(props, context);

    let {tool, value, width: viewerWidth, height: viewerHeight, scaleFactorMin, scaleFactorMax, children} = this.props;
    let {width: SVGWidth, height: SVGHeight} = children.props;

    this.state = {
      value: value ? value : getDefaultValue(viewerWidth, viewerHeight, SVGWidth, SVGHeight, scaleFactorMin, scaleFactorMax),
      tool: tool ? tool : TOOL_NONE
    };
    this.ViewerDOM = null;

    this.autoPanLoop = this.autoPanLoop.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    let value = this.getValue();
    let needUpdate = false;
    let nextValue = value;

    if (value.viewerWidth !== nextProps.width || value.viewerHeight !== nextProps.height) {
      nextValue = setViewerSize(nextValue, nextProps.width, nextProps.height);
      needUpdate = true;
    }

    let {width: SVGWidth, height: SVGHeight} = nextProps.children.props;
    if (value.SVGWidth !== SVGWidth || value.SVGHeight !== SVGHeight) {
      nextValue = setSVGSize(nextValue, SVGWidth, SVGHeight);
      needUpdate = true;
    }

    if (value.scaleFactorMin !== nextProps.scaleFactorMin || value.scaleFactorMax !== nextProps.scaleFactorMax) {
      nextValue = setZoomLevels(nextValue, nextProps.scaleFactorMin, nextProps.scaleFactorMax);
      needUpdate = true;
    }

    if (needUpdate) {
      this.setValue(nextValue);
    }
  }


  getValue() {
    return this.props.value ? this.props.value : this.state.value;
  }

  getTool() {
    return this.props.tool ? this.props.tool : this.state.tool;
  }

  getSvgStyle(cursor) {
    const style = {display: 'block'};

    if (cursor) {
      style.cursor = cursor;
    }

    if (this.props.detectPinchGesture || [TOOL_PAN, TOOL_AUTO].indexOf(this.getTool()) !== -1) {
      style.touchAction = 'none';
    }

    return style;
  }

  setValue(nextValue) {
    let {onChangeValue, onZoom, onPan} = this.props;
    this.setState({value: nextValue});
    if (onChangeValue) onChangeValue(nextValue);
    if (nextValue.lastAction) {
      if (onZoom && nextValue.lastAction === ACTION_ZOOM) onZoom(nextValue);
      if (onPan && nextValue.lastAction === ACTION_PAN) onPan(nextValue);
    }
  }

  pan(SVGDeltaX, SVGDeltaY) {
    let nextValue = pan(this.getValue(), SVGDeltaX, SVGDeltaY);
    this.setValue(nextValue);
  }

  zoom(SVGPointX, SVGPointY, scaleFactor) {
    let nextValue = zoom(this.getValue(), SVGPointX, SVGPointY, scaleFactor);
    this.setValue(nextValue);
  }

  fitSelection(selectionSVGPointX, selectionSVGPointY, selectionWidth, selectionHeight) {
    let nextValue = fitSelection(this.getValue(), selectionSVGPointX, selectionSVGPointY, selectionWidth, selectionHeight);
    this.setValue(nextValue);
  }

  fitToViewer(SVGAlignX=ALIGN_LEFT, SVGAlignY=ALIGN_TOP) {
    let nextValue = fitToViewer(this.getValue(), SVGAlignX, SVGAlignY);
    this.setValue(nextValue);
  }

  zoomOnViewerCenter(scaleFactor) {
    let nextValue = zoomOnViewerCenter(this.getValue(), scaleFactor);
    this.setValue(nextValue);
  }

  setPointOnViewerCenter(SVGPointX, SVGPointY, zoomLevel) {
    let nextValue = setPointOnViewerCenter(this.getValue(), SVGPointX, SVGPointY, zoomLevel);
    this.setValue(nextValue);
  }

  reset() {
    let nextValue = reset(this.getValue());
    this.setValue(nextValue);
  }

  changeTool(tool) {
    this.setState({tool});
    if (this.props.onChangeTool) this.props.onChangeTool(tool);
  }

  openMiniature() {
    let nextValue = openMiniature(this.getValue());
    this.setValue(nextValue);
  }

  closeMiniature() {
    let nextValue = closeMiniature(this.getValue());
    this.setValue(nextValue);
  }

  handleViewerEvent(event) {
    let {props, state: {value}, ViewerDOM} = this;

    if (!([TOOL_NONE, TOOL_AUTO].indexOf(this.getTool()) >= 0)) return;
    if (event.target === ViewerDOM) return;

    let eventsHandler = {
      click: props.onClick,
      dblclick: props.onDoubleClick,

      mousemove: props.onMouseMove,
      mouseup: props.onMouseUp,
      mousedown: props.onMouseDown,

      touchstart: props.onTouchStart,
      touchmove: props.onTouchMove,
      touchend: props.onTouchEnd,
      touchcancel: props.onTouchCancel,
    };

    let onEventHandler = eventsHandler[event.type];
    if (!onEventHandler) return;

    onEventHandler(eventFactory(event, value, ViewerDOM));
  }

  autoPanLoop() {
    let coords = {x: this.state.viewerX, y: this.state.viewerY};
    let nextValue = onInterval(null, this.ViewerDOM, this.getTool(), this.getValue(), this.props, coords);

    if (this.getValue() !== nextValue) {
      this.setValue(nextValue);
    }

    if (this.autoPanIsRunning) {
      requestAnimationFrame(this.autoPanLoop);
    }
  }

  componentDidMount() {
    let {props, state} = this;
    if (props.onChangeValue) props.onChangeValue(state.value);

    this.autoPanIsRunning = true;
    requestAnimationFrame(this.autoPanLoop);
  }

  componentWillUnmount() {
    this.autoPanIsRunning = false;
  }

  render() {
    let {props, state: {viewerX, viewerY}} = this;
    let tool = this.getTool();
    let value = this.getValue();
    let {customToolbar: CustomToolbar, customMiniature: CustomMiniature} = props;

    let panningWithToolAuto = tool === TOOL_AUTO
      && value.mode === MODE_PANNING
      && value.startX !== value.endX
      && value.startY !== value.endY;

    let cursor;

    if (tool === TOOL_PAN)
      cursor = cursorPolyfill(value.mode === MODE_PANNING ? 'grabbing' : 'grab');

    if (tool === TOOL_ZOOM_IN)
      cursor = cursorPolyfill('zoom-in');

    if (tool === TOOL_ZOOM_OUT)
      cursor = cursorPolyfill('zoom-out');

    if (panningWithToolAuto)
      cursor = cursorPolyfill('grabbing');

    let blockChildEvents = [TOOL_PAN, TOOL_ZOOM_IN, TOOL_ZOOM_OUT].indexOf(tool) >= 0;
    blockChildEvents = blockChildEvents || panningWithToolAuto;

    return (
      <div
        style={{position: "relative", width: value.viewerWidth, height: value.viewerHeight, ...props.style}}
        className={this.props.className}>
        <svg
          ref={ViewerDOM => this.ViewerDOM = ViewerDOM}
          width={value.viewerWidth}
          height={value.viewerHeight}
          style={this.getSvgStyle(cursor)}

          onMouseDown={event => {
            let nextValue = onMouseDown(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props);
            if (this.getValue() !== nextValue) this.setValue(nextValue);
            this.handleViewerEvent(event);
          }}
          onMouseMove={event => {
            let {left, top} = this.ViewerDOM.getBoundingClientRect();
            let x = event.clientX - Math.round(left);
            let y = event.clientY - Math.round(top);

            let nextValue = onMouseMove(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props, {x, y});
            if (this.getValue() !== nextValue) this.setValue(nextValue);
            this.setState({viewerX: x, viewerY: y});
            this.handleViewerEvent(event);
          }}
          onMouseUp={event => {
            let nextValue = onMouseUp(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props);
            if (this.getValue() !== nextValue) this.setValue(nextValue);
            this.handleViewerEvent(event);
          }}

          onClick={event => {
            this.handleViewerEvent(event)
          }}
          onDoubleClick={event => {
            let nextValue = onDoubleClick(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props);
            if (this.getValue() !== nextValue) this.setValue(nextValue);
            this.handleViewerEvent(event);
          }}

          onWheel={event => {
            let nextValue = onWheel(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props);
            if (this.getValue() !== nextValue) this.setValue(nextValue);
          }}

          onMouseEnter={event => {
            if (detectTouch()) return;
            let nextValue = onMouseEnterOrLeave(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props);
            if (this.getValue() !== nextValue) this.setValue(nextValue);
          }}
          onMouseLeave={event => {
            let nextValue = onMouseEnterOrLeave(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props);
            if (this.getValue() !== nextValue) this.setValue(nextValue);
          }}

          onTouchStart={event => {
            let nextValue = onTouchStart(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props);
            if (this.getValue() !== nextValue) this.setValue(nextValue);
            this.handleViewerEvent(event);
          }}
          onTouchMove={event => {
            let nextValue = onTouchMove(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props);
            if (this.getValue() !== nextValue) this.setValue(nextValue);
            this.handleViewerEvent(event);
          }}
          onTouchEnd={event => {
            let nextValue = onTouchEnd(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props);
            if (this.getValue() !== nextValue) this.setValue(nextValue);
            this.handleViewerEvent(event);
          }}
          onTouchCancel={event => {
            let nextValue = onTouchCancel(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props);
            if (this.getValue() !== nextValue) this.setValue(nextValue);
            this.handleViewerEvent(event);
          }}>

          <rect
            fill={props.background}
            x={0}
            y={0}
            width={value.viewerWidth}
            height={value.viewerHeight}
            style={{pointerEvents: "none"}}
          />

          <g
            transform={toSVG(value)}
            style={blockChildEvents ? {pointerEvents: "none"} : {}}>
            <rect
              fill={this.props.SVGBackground}
              style={this.props.SVGStyle}
              x={0}
              y={0}
              width={value.SVGWidth}
              height={value.SVGHeight}/>
            <g>
              {props.children.props.children}
            </g>
          </g>

          {!([TOOL_NONE, TOOL_AUTO].indexOf(tool) >= 0 && props.detectAutoPan && value.focus) ? null : (
            <g style={{pointerEvents: "none"}}>
              {!(viewerY <= 20) ? null :
                <BorderGradient direction={POSITION_TOP} width={value.viewerWidth} height={value.viewerHeight}/>
              }

              {!(value.viewerWidth - viewerX <= 20) ? null :
                <BorderGradient direction={POSITION_RIGHT} width={value.viewerWidth} height={value.viewerHeight}/>
              }

              {!(value.viewerHeight - viewerY <= 20) ? null :
                <BorderGradient direction={POSITION_BOTTOM} width={value.viewerWidth} height={value.viewerHeight}/>
              }

              {!(value.focus && viewerX <= 20) ? null :
                <BorderGradient direction={POSITION_LEFT} width={value.viewerWidth} height={value.viewerHeight}/>
              }
            </g>
          )}

          {!(value.mode === MODE_ZOOMING) ? null :
            <Selection startX={value.startX} startY={value.startY} endX={value.endX} endY={value.endY}/>
          }
        </svg>

        {props.toolbarPosition === POSITION_NONE ? null :
          <CustomToolbar
            position={props.toolbarPosition}
            value={value}
            onChangeValue={value => this.setValue(value)}
            tool={tool}
            onChangeTool={tool => this.changeTool(tool)}
            {...this.props.toolbarProps} />}

        {props.miniaturePosition === POSITION_NONE ? null :
          <CustomMiniature
            position={props.miniaturePosition}
            value={value}
            onChangeValue={value => this.setValue(value)}
            SVGBackground={this.props.SVGBackground}
            background={this.props.miniatureBackground}
            width={this.props.miniatureWidth}
            height={this.props.miniatureHeight}
          >
            {props.children.props.children}
          </CustomMiniature>
        }
      </div>
    );
  }
}

ReactSVGPanZoom.propTypes = {
  //width of the viewer displayed on screen
  width: PropTypes.number.isRequired,

  //height of the viewer displayed on screen
  height: PropTypes.number.isRequired,

  //background of the viewer
  background: PropTypes.string,

  //background of the svg
  SVGBackground: PropTypes.string,

  //style of the svg
  SVGStyle: PropTypes.object,

  //value of the viewer (current point of view)
  value: PropTypes.shape({
    version: PropTypes.oneOf([2]).isRequired,
    mode: PropTypes.oneOf([MODE_IDLE, MODE_PANNING, MODE_ZOOMING]).isRequired,
    focus: PropTypes.bool.isRequired,
    a: PropTypes.number.isRequired,
    b: PropTypes.number.isRequired,
    c: PropTypes.number.isRequired,
    d: PropTypes.number.isRequired,
    e: PropTypes.number.isRequired,
    f: PropTypes.number.isRequired,
    viewerWidth: PropTypes.number.isRequired,
    viewerHeight: PropTypes.number.isRequired,
    SVGWidth: PropTypes.number.isRequired,
    SVGHeight: PropTypes.number.isRequired,
    startX: PropTypes.number,
    startY: PropTypes.number,
    endX: PropTypes.number,
    endY: PropTypes.number,
    miniatureOpen: PropTypes.bool.isRequired,
  }),

  //CSS style of the Viewer
  style: PropTypes.object,

  //className of the Viewer
  className: PropTypes.string,

  //perform zoom operation on mouse scroll
  detectWheel: PropTypes.bool,

  //perform PAN if the mouse is on viewer border
  detectAutoPan: PropTypes.bool,

  //perform zoom operation on pinch gesture
  detectPinchGesture: PropTypes.bool,

  //toolbar position
  toolbarPosition: PropTypes.oneOf([POSITION_NONE, POSITION_TOP, POSITION_RIGHT, POSITION_BOTTOM, POSITION_LEFT]),

  //handler something changed
  onChangeValue: PropTypes.func,

  //handler tool changed
  onChangeTool: PropTypes.func,

  //handler zoom level changed
  onZoom: PropTypes.func,

  //handler pan action performed
  onPan: PropTypes.func,

  //handler click
  onClick: PropTypes.func,

  //handler double click
  onDoubleClick: PropTypes.func,

  //handler mouseup
  onMouseUp: PropTypes.func,

  //handler mousemove
  onMouseMove: PropTypes.func,

  //handler mousedown
  onMouseDown: PropTypes.func,

  //if disabled the user can move the image outside the viewer
  preventPanOutside: PropTypes.bool,

  //how much scale in or out
  scaleFactor: PropTypes.number,

  //how much scale in or out on mouse wheel (requires detectWheel enabled)
  scaleFactorOnWheel: PropTypes.number,

  // maximum amount of scale a user can zoom in to
  scaleFactorMax: PropTypes.number,

  // minimum amount of a scale a user can zoom out of
  scaleFactorMin: PropTypes.number,

  //current active tool (TOOL_NONE, TOOL_PAN, TOOL_ZOOM_IN, TOOL_ZOOM_OUT)
  tool: PropTypes.oneOf([TOOL_AUTO, TOOL_NONE, TOOL_PAN, TOOL_ZOOM_IN, TOOL_ZOOM_OUT]),

  //modifier keys //https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/getModifierState
  modifierKeys: PropTypes.array,

  //override toolbar component
  customToolbar: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),

  //miniature position
  miniaturePosition: PropTypes.oneOf([POSITION_NONE, POSITION_RIGHT, POSITION_LEFT]),

  //miniature height
  miniatureBackground: PropTypes.string,

  //miniature width
  miniatureWidth: PropTypes.number,

  //miniature height
  miniatureHeight: PropTypes.number,

  //override miniature component
  customMiniature: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),

  //Turn off zoom on double click
  disableDoubleClickZoomWithToolAuto: PropTypes.bool,

  //toolbar props
  toolbarProps: PropTypes.shape({
    SVGAlignX: PropTypes.oneOf([ALIGN_CENTER, ALIGN_LEFT, ALIGN_RIGHT]),
    SVGAlignY: PropTypes.oneOf([ALIGN_CENTER, ALIGN_TOP, ALIGN_BOTTOM]),
  }),

  //accept only one node SVG
  children: function (props, propName, componentName) {
    // Only accept a single child, of the appropriate type
    //credits: http://www.mattzabriskie.com/blog/react-validating-children
    let prop = props[propName];
    let types = ['svg'];
    if (React.Children.count(prop) !== 1 ||
      types.indexOf(prop.type) === -1) {
      return new Error(
        '`' + componentName + '` ' +
        'should have a single child of the following types: ' +
        ' `' + types.join('`, `') + '`.'
      );
    }
    if (!prop.props.hasOwnProperty('width') || !prop.props.hasOwnProperty('height')) {
      return new Error('SVG should have props `width` and `height`');
    }

  }
};

ReactSVGPanZoom.defaultProps = {
  value: null,
  tool: null,
  style: {},
  background: "#616264",
  SVGBackground: "#fff",
  SVGStyle: {},
  detectWheel: true,
  detectAutoPan: true,
  detectPinchGesture: true,
  toolbarPosition: POSITION_RIGHT,
  modifierKeys: ["Alt", "Shift", "Control"],
  customToolbar: Toolbar,
  preventPanOutside: true,
  scaleFactor: 1.1,
  scaleFactorOnWheel: 1.06,
  miniaturePosition: POSITION_LEFT,
  miniatureWidth: 100,
  miniatureHeight: 80,
  miniatureBackground: "#616264",
  customMiniature: Miniature,
  disableZoomWithToolAuto: false,
  onZoom: null,
  onPan: null,
  toolbarProps: {}
};
