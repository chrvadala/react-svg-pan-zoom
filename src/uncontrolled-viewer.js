import React from "react";
import ReactSVGPanZoom from './viewer'
import PropTypes from "prop-types";
import {TOOL_NONE} from "./constants";

export default class UncontrolledReactSVGPanZoom extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      value: props.defaultValue || {},
      tool: props.defaultTool || TOOL_NONE,
    }

    this.Viewer = null;
    this.changeTool = this.changeTool.bind(this)
    this.changeValue = this.changeValue.bind(this)
  }

  changeTool(tool) {
    this.setState({tool})
  }

  changeValue(value) {
    this.setState({value})
  }

  pan(SVGDeltaX, SVGDeltaY) {
    this.Viewer.pan(SVGDeltaX, SVGDeltaY)
  }

  zoom(SVGPointX, SVGPointY, scaleFactor) {
    this.Viewer.zoom(SVGPointX, SVGPointY, scaleFactor)
  }

  fitSelection(selectionSVGPointX, selectionSVGPointY, selectionWidth, selectionHeight, SVGAlignX = ALIGN_LEFT, SVGAlignY = ALIGN_TOP) {
    this.Viewer.fitSelection(selectionSVGPointX, selectionSVGPointY, selectionWidth, selectionHeight, SVGAlignX, SVGAlignY)
  }

  fitToViewer(SVGAlignX = ALIGN_LEFT, SVGAlignY = ALIGN_TOP) {
    this.Viewer.fitToViewer(SVGAlignX, SVGAlignY)
  }

  zoomOnViewerCenter(scaleFactor) {
    this.Viewer.zoomOnViewerCenter(scaleFactor)
  }

  setPointOnViewerCenter(SVGPointX, SVGPointY, zoomLevel) {
    this.Viewer.setPointOnViewerCenter(SVGPointX, SVGPointY, zoomLevel)
  }

  reset() {
    this.Viewer.reset()
  }

  openMiniature() {
    this.Viewer.openMiniature()
  }

  closeMiniature() {
    this.Viewer.closeMiniature()
  }

  render() {
    const {width, height, onChangeTool, onChangeValue, ...props} = this.props
    const {tool, value} = this.state

    return (
      <ReactSVGPanZoom
        width={width} height={height}
        tool={tool} onChangeTool={this.changeTool}
        value={value} onChangeValue={this.changeValue}
        ref={Viewer => this.Viewer = Viewer}
        {...props}
      />
    )
  }
}

UncontrolledReactSVGPanZoom.propTypes = {
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,

  defaultValue: PropTypes.object,
  defaultTool: PropTypes.string,
}
