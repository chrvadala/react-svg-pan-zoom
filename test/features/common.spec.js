import {
  decompose,
  getDefaultValue,
  getSVGPoint,
  isValueValid,
  reset,
  resetMode,
  set,
  setFocus,
  setPointOnViewerCenter,
  setSVGViewBox,
  setViewerSize,
  setZoomLevels
} from "../../src/features/common";
import {fitToViewer, MODE_IDLE, pan} from "../../src";

const VALUE = getDefaultValue(
  200, 200, //viewer 200x200
  0, 0, 400, 400, //svg 400x400
)


test("getDefaultValue", () => {
  const value = getDefaultValue(
    10, 20,
    5, 6, 7, 8,
    80, 90
  )

  expect(value.mode).toBe(MODE_IDLE)

  expect(value.viewerWidth).toBe(10)
  expect(value.viewerHeight).toBe(20)

  expect(value.SVGMinX).toBe(5)
  expect(value.SVGMinY).toBe(6)
  expect(value.SVGWidth).toBe(7)
  expect(value.SVGHeight).toBe(8)

  expect(value.scaleFactorMin).toBe(80)
  expect(value.scaleFactorMax).toBe(90)
})

test("set", () => {
  const value = getDefaultValue()

  const nextValue = set(value, {a: 1, b: 2, c: 3, d: 4, e: 6}, 'ZOOM')

  expect(nextValue).not.toBe(value)
  expect(nextValue).toMatchObject({a: 1, b: 2, c: 3, d: 4, e: 6})
  expect(function () {
    nextValue.a = 100
  }).toThrow()
})

test("isValueValid", () => {
  expect(isValueValid({version: 99})).toBe(false)
  expect(isValueValid('123456789')).toBe(false)

  expect(isValueValid({version: 3})).toBe(true)
})

test("getSVGPoint", () => {
  expect(getSVGPoint(VALUE, 100, 100)).toMatchObject({x: 100, y: 100})
  expect(getSVGPoint(VALUE, 200, 200)).toMatchObject({x: 200, y: 200})

  const value_2 = fitToViewer(VALUE)
  expect(getSVGPoint(value_2, 100, 100)).toMatchObject({x: 200, y: 200})
})

test("decompose", () => {
  expect(decompose(VALUE)).toMatchObject({scaleFactor: 1, translationX: 0, translationY: 0})

  const value_2 = fitToViewer(VALUE)
  expect(decompose(value_2)).toMatchObject({scaleFactor: 0.5, translationX: 0, translationY: 0})

  const value_3 = pan(VALUE, 100, 200)
  expect(decompose(value_3)).toMatchObject({scaleFactor: 1, translationX: 100, translationY: 200})

  const value_4 = pan(fitToViewer(VALUE), 100, 200)
  expect(decompose(value_4)).toMatchObject({scaleFactor: 0.5, translationX: 50, translationY: 100})
})

test("setFocus", () => {
  expect(VALUE).toMatchObject({focus: false})
  const value_2 = setFocus(VALUE, true)
  expect(value_2).toMatchObject({focus: true})
})

test("setViewerSize", () => {
  expect(VALUE).toMatchObject({viewerWidth: 200, viewerHeight: 200})
  const value_2 = setViewerSize(VALUE, 50, 100)
  expect(value_2).toMatchObject({viewerWidth: 50, viewerHeight: 100})
})

test("setSVGViewBox", () => {
  expect(VALUE).toMatchObject({SVGMinX: 0, SVGMinY: 0, SVGWidth: 400, SVGHeight: 400})
  const value_2 = setSVGViewBox(VALUE, 30, 40, 130, 140)
  expect(value_2).toMatchObject({SVGMinX: 30, SVGMinY: 40, SVGWidth: 130, SVGHeight: 140})
})

test("setZoomLevels", () => {
  expect(VALUE).toMatchObject({scaleFactorMin: null, scaleFactorMax: null})
  const value_2 = setZoomLevels(VALUE, 0.1, 99)
  expect(value_2).toMatchObject({scaleFactorMin: 0.1, scaleFactorMax: 99})
})

test("setPointOnViewerCenter", () => {
  const value_2 = setPointOnViewerCenter(VALUE, 200, 200, 1) //move SVG center to viewer center
  expect(value_2).toMatchSnapshot()
  expect(decompose(value_2)).toMatchObject({translationX: -100, translationY: -100})

  const value_3 = setPointOnViewerCenter(VALUE, 200, 200, 2) //move SVG center to viewer center and apply zoom
  expect(value_3).toMatchSnapshot()
  const translation = (((400 * 2) - 200) / 2) // (((SVGWidth * zoomLevel) - viewerWidth) / 2)
  expect(decompose(value_3)).toMatchObject({translationX: -translation, translationY: -translation})
})

test("reset", () => {
  const value_2 = fitToViewer(VALUE) //apply something
  expect(value_2).not.toBe(VALUE)
  expect(value_2).not.toMatchObject(VALUE)

  const value_3 = reset(value_2)
  expect(value_3).not.toBe(VALUE)
  expect(value_3).toMatchObject(VALUE)
})

test("resetMode", () => {
  const value_2 = fitToViewer(VALUE) //apply something
  const value_3 = set(value_2, {mode: 'ABC', startX: 10, startY: 20, endX: 100, endY: 200})

  expect(value_3).toMatchObject({mode: 'ABC', startX: 10, startY: 20, endX: 100, endY: 200})
  const value_4 = resetMode(value_3)
  expect(value_4).toMatchObject({mode: 'idle', startX: null, startY: null, endX: null, endY: null})
})
