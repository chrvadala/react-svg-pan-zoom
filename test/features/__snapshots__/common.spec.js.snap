// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setPointOnViewerCenter 1`] = `
Object {
  "SVGHeight": 400,
  "SVGMinX": 0,
  "SVGMinY": 0,
  "SVGWidth": 400,
  "a": 1,
  "b": 0,
  "c": 0,
  "d": 1,
  "e": -100,
  "endX": null,
  "endY": null,
  "f": -100,
  "focus": false,
  "lastAction": null,
  "miniatureOpen": true,
  "mode": "idle",
  "pinchPointDistance": null,
  "prePinchMode": null,
  "scaleFactorMax": null,
  "scaleFactorMin": null,
  "startX": null,
  "startY": null,
  "version": 3,
  "viewerHeight": 200,
  "viewerWidth": 200,
}
`;

exports[`setPointOnViewerCenter 2`] = `
Object {
  "SVGHeight": 400,
  "SVGMinX": 0,
  "SVGMinY": 0,
  "SVGWidth": 400,
  "a": 2,
  "b": 0,
  "c": 0,
  "d": 2,
  "e": -300,
  "endX": null,
  "endY": null,
  "f": -300,
  "focus": false,
  "lastAction": null,
  "miniatureOpen": true,
  "mode": "idle",
  "pinchPointDistance": null,
  "prePinchMode": null,
  "scaleFactorMax": null,
  "scaleFactorMin": null,
  "startX": null,
  "startY": null,
  "version": 3,
  "viewerHeight": 200,
  "viewerWidth": 200,
}
`;
